language/compiler/bytecode_source_map/src/disassembler.rs

```diff
@@ -76,6 +85,145 @@ impl<Location: Clone + Eq + Default> Disassembler<Location> {
+
+    fn name_for_field(&self, field_idx: FieldDefinitionIndex) -> Result<(String, String)> {
+        let field_def = self.source_mapper.bytecode.field_def_at(field_idx);
+        let struct_def_idx = self
+            .source_mapper
+            .bytecode
+            .struct_defs()
+            .iter()
+            .position(|struct_def| struct_def.struct_handle == field_def.struct_)
+            .ok_or_else(|| format_err!("Unable to find struct definition for _BorrowField"))?;
+        let field_name = self
+            .source_mapper
+            .bytecode
+            .identifier_at(field_def.name)
+            .to_string();
+        let struct_def = self
+            .source_mapper
+            .bytecode
+            .struct_def_at(StructDefinitionIndex(struct_def_idx as TableIndex));
+        let struct_handle = self
+            .source_mapper
+            .bytecode
+            .struct_handle_at(struct_def.struct_handle);
+        let struct_name = self
+            .source_mapper
+            .bytecode
+            .identifier_at(struct_handle.name)
+            .to_string();
+        let qualified_name = format!("{}.{}", struct_name, field_name);
+        let struct_source_info = self
+            .source_mapper
+            .source_map
+            .get_struct_source_map(StructDefinitionIndex(struct_def_idx as TableIndex))?;
+        let field_type_sig = self
+            .source_mapper
+            .bytecode
+            .type_signature_at(field_def.signature);
+        let ty = self.disassemble_sig_tok(
+            field_type_sig.0.clone(),
+            &struct_source_info.type_parameters,
+        )?;
+        Ok((qualified_name, ty))
+    }
+
+    fn struct_ty_info(
+        &self,
+        struct_idx: &StructDefinitionIndex,
+        types_idx: &LocalsSignatureIndex,
+    ) -> Result<String> {
+        let struct_definition = self.get_struct_def(*struct_idx)?;
+        let struct_source_map = self
+            .source_mapper
+            .source_map
+            .get_struct_source_map(*struct_idx)?;
+        let locals_signature = self.source_mapper.bytecode.locals_signature_at(*types_idx);
+        let type_arguments = locals_signature
+            .0
+            .iter()
+            .map(|sig_tok| {
+                Ok(self.disassemble_sig_tok(sig_tok.clone(), &struct_source_map.type_parameters)?)
+            })
+            .collect::<Result<Vec<String>>>()?;
+
+        let struct_handle = self
+            .source_mapper
+            .bytecode
+            .struct_handle_at(struct_definition.struct_handle);
+        let name = self
+            .source_mapper
+            .bytecode
+            .identifier_at(struct_handle.name)
+            .to_string();
+        Ok(format!(
+            "{idx}, {name}{ty_instantiation}",
+            idx = struct_idx,
+            name = name,
+            ty_instantiation = Self::enclose_ty_params(&type_arguments)
```

Comment:
The returned artifact here seems a bit odd:

- The caller passes the `struct_idx`, but it also occurs in the printed output
- It might be useful to return the name and type parameters separately in case the caller wants to do different things with each.

Reply:
I agree on the generality argument. I think splitting these two off into separate functions is a good idea, and doing that forces the answer to your first point (no longer doing this)